petab/v2/lint.py

```diff
@@ -0,0 +1,550 @@
+                problem.mapping_df,
+            )
+        except AssertionError as e:
+            return ValidationIssue(
+                level=ValidationIssueSeverity.ERROR, message=str(e)
+            )
+
+
+class CheckVisualizationTable(ValidationTask):
+    """A task to validate the visualization table of a PEtab problem."""
+
+    def run(self, problem: Problem) -> ValidationIssue | None:
+        if problem.visualization_df is None:
+            return
+
+        from petab.visualize.lint import validate_visualization_df
+
+        if validate_visualization_df(problem):
+            return ValidationIssue(
+                level=ValidationIssueSeverity.ERROR,
+                message="Visualization table is invalid.",
+            )
+
+
+def get_required_parameters_for_parameter_table(
+    problem: Problem,
+) -> set[str]:
+    """
+    Get set of parameters which need to go into the parameter table
+
+    Arguments:
+        problem: The PEtab problem
+    Returns:
+        Set of parameter IDs which PEtab requires to be present in the
+        parameter table. That is all {observable,noise}Parameters from the
+        measurement table as well as all parametric condition table overrides
+        that are not defined in the model.
+    """
+
+    # use ordered dict as proxy for ordered set
+    parameter_ids = OrderedDict()
+
+    # Add parameters from measurement table, unless they are fixed parameters
+    def append_overrides(overrides):
+        for p in overrides:
+            if isinstance(p, str) and p not in problem.condition_df.columns:
+                parameter_ids[p] = None
+
+    for _, row in problem.measurement_df.iterrows():
+        # we trust that the number of overrides matches
+        append_overrides(
+            split_parameter_replacement_list(
+                row.get(OBSERVABLE_PARAMETERS, None)
+            )
+        )
+        append_overrides(
+            split_parameter_replacement_list(row.get(NOISE_PARAMETERS, None))
+        )
+
+    # Add output parameters except for placeholders
+    for formula_type, placeholder_sources in (
+        (
+            # Observable formulae
+            {"observables": True, "noise": False},
+            # can only contain observable placeholders
+            {"noise": False, "observables": True},
+        ),
+        (
+            # Noise formulae
+            {"observables": False, "noise": True},
+            # can contain noise and observable placeholders
+            {"noise": True, "observables": True},
+        ),
+    ):
+        output_parameters = get_output_parameters(
+            problem.observable_df,
+            problem.model,
+            mapping_df=problem.mapping_df,
+            **formula_type,
+        )
+        placeholders = get_placeholders(
+            problem.observable_df,
+            **placeholder_sources,
+        )
+        for p in output_parameters:
+            if p not in placeholders and p not in problem.observable_df.index:
```

Comment:
```suggestion
            # remove `p not in problem.observable_df.index` when
            # `get_output_parameters` is updated for PEtab v2/v1.1, where
            # observable IDs are allowed in observable formulae
            if p not in placeholders and p not in problem.observable_df.index:
```

If `parameter_ids` was a set, we could do `parameter_ids |= output_parameters - placeholders`. We could `return sorted(parameter_ids)` to simplify the code a bit. The `dict` makes the code a little weird but fine as is.

Reply:
I don't recall the reason for insisting on the given order, but I don't see a problem with requiring `sorted` for deterministic results. Also currently the type annotation is wrong. Will change.